Add top-level access to AesGcm encoding scheme

tiesselune opened the pull request:
As discussed in #35, the current API does not allow access to encryption and decryption functionality for the `AesGcm` legacy encoding scheme without implementing a new back-end, even though it *is* fully implemented for OpenSSL internally. 

This pull request provides an access to the AesGcm encoding scheme through top-level functions in a `legacy` module.

It tries to mimic the main `Aes128Gcm` top-level library structure and tests.

So far, 4 tests have been implemented, using the IETF example data from [IETF Web Push Encryption Draft 4](https://tools.ietf.org/html/draft-ietf-webpush-encryption-04#appendix-A), with no padding and an `rs` of 4096.

For this to work, a minor fix has been implemented in the `AesGcmEceWebPush::encrypt_with_keys` function to allow the use of the salt provided in the `params` argument, which was previously ignored and made the tests fail.

To finalize this pull request, it would be better to provide more tests but for the moment I am unsure of where to get or generate relevant test data.

rfk commented:
@jrconlin could you please gut-check the new API that's being proposed here, particularly that the "legacy" naming is conveying approximate the right thing? (It seems right to me, but you've deeper experience with the history of these crypto methods).

rfk commented on src/legacy.rs:
```diff
@@ -0,0 +1,194 @@
+pub fn encrypt_aesgcm(
+    remote_pub: &[u8],
+    remote_auth: &[u8],
+    salt: &[u8],
```
I think ideally we would not accept a `salt` argument here, since it just creates an opportunity for the caller to do something wrong when generating it. However, the existing top-level `encrypt` function currently takes a `salt` argument, so it makes sense to mirror that here 👍 

I'm mostly saying this out loud as a reminder to myself that a future API change should remove `salt` from both of these functions...

tiesselune replied:
Well it does provide a nice way of testing the function result against the IETF versions.
If removed, those tests would need to be done on a lower-level layer, possibly recoding part of the function.

A suggestion I could make would be to have private versions of these functions, with the `salt` argument, that could be tested, and a public version that just generates a random salt and calls the underlying testable and private function ... ?

rfk replied:
Indeed, it's very convenient for the tests. Figuring out how to do better here can be a problem for our future selves.

rfk commented on src/legacy.rs:
```diff
@@ -0,0 +1,194 @@
+
+/// Decrypt a block using legacy AESGCM encoding.
+///
+/// param components &str - The locally generated private key components.
```
The `components` argument should be documented as an `&EcKeyComponents`, not a `&str`. (I notice that the existing `decrypt` function has the same problem, unfortunately; given that rustdoc generates really nice automated function signature references, maybe we should just remove this narrative "param foo &bar" documentation entirely?)

tiesselune replied:
That's right, I copied and pasted the comment but did not pay attention to the problem. I'll add a fixing commit.

rfk commented on src/legacy.rs:
```diff
@@ -0,0 +1,194 @@
+        let decrypted =
+            AesGcmEceWebPush::decrypt(&*remote_key, &auth_secret, &encrypted_block).unwrap();
+        assert_eq!(decrypted, plaintext.to_vec());
+    }
```
This test is very similar to the one in [the main `lib.rs`](https://github.com/mozilla/rust-ece/blob/64af2968ebaab43b5a2206795ad64de3153e585f/src/lib.rs#L410), except that your one uses an explicit salt. What do you think about replacing the one linked above with your version here rather than having two nearly-identical tests?

rfk commented on src/legacy.rs:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    #[test]
+    fn try_encrypt_ietf_rfc() {
```
To cross-check, are these the same values from [the example in Section 5](https://tools.ietf.org/html/draft-ietf-webpush-encryption-04#section-5) of the linked RFC? It would be great if you could include a link in a comment here right the code for future readers to be able to find the source material.

tiesselune replied:
I took the example from draft 4, but they are indeed the same. I have just converted URL-safe Base64 to hexadecimal. It is doable for sure. I'll submit a few commits to handle those comments later today ;)

rfk reviewed (changes requested):
This looks great, thank you! I left a couple of small requests below just about organization of the code, but I'm confident we can land this once you're happy with it.

> To finalize this pull request, it would be better to provide more tests but for the moment
> I am unsure of where to get or generate relevant test data.

Honestly, I think the test coverage you've added is sufficient here considering the change you're making - it shouldn't have to be your job to add a bunch of tests for the existing `aesgcm` code in the crate, so I wouldn't hesitate to merge this as long as the new functions you're adding have coverage tested.

More broadly, it is true that the `aesgcm` code itself could use some better test coverage, although I'm not sure if you saw the small set of existing tests in `lib.rs`. For something like this I would typically try to get a known-good implementation to generate fixed test vectors, perhaps by scripting up the [node web-push package](https://www.npmjs.com/package/web-push) to encrypt some known strings and testing that the Rust code here can successfully decrypt them. But I will understand if that's a lot more than you're willing to take on just now :-)

tiesselune commented:
I have made a few fix commits to handle most of the issues addressed up there.
I've removed the redundant test and added a disclaimer in the documentation for the AesGcm and Aes128Gcm encryption structs (since they appear higher in the documentation than the `encrypt` and `decrypt` functions, and tend to attract clicks) to redirect to the `encrypt` and `decrypt` functions of both schemes.

For the test part, I've taken a look at what's possible and the problem with the node [`web-push`](https://www.npmjs.com/package/web-push) library or the [`http_ece`](https://www.npmjs.com/package/http_ece) one is that they don't expose an API that allows for different RS or padding length parameters... So I can't mimic the main library's tests without spending a lot more time taking it apart to generate test data. I've taken a look at the [`ecec`](https://github.com/web-push-libs/ecec) library as well but it comes with the certainty that working with C is going to be even more cumbersome 😅

Anyway I'm going to get this Pull Request out of draft state, at least, and might open another issue to handle the test issue, which is something I should be able to tackle a bit later if you still need the workpower.

rfk commented on src/aes128gcm.rs:
```diff
@@ -26,6 +26,9 @@ const ECE_AES128GCM_NONCE_INFO: &str = "Content-Encoding: nonce\0";
 
+/// Web Push encryption structure for the AES128GCM encoding scheme ([RFC8591](https://tools.ietf.org/html/rfc8291))
+/// 
+/// This structure is meant for advanced use. For simple encryption/decryption, use the top-level [`encrypt`](crate::encrypt) and [`decrypt`](crate::decrypt) functions.
```
Great, thank you!

rfk commented on src/lib.rs:
```diff
@@ -29,11 +30,12 @@ pub fn generate_keypair_and_auth_secret(
+/// * `remote_pub` : The remote public key
+/// * `remote_auth` : The remote authorization token
+/// * `salt` : The locally generated random salt
+/// * `data` : The data to encrypt
```
Oh nice, thanks for tackling these here as well.

rfk reviewed (approved):
This looks excellent, I'll merge and see about getting it into a release.

rfk commented:
I've published `v1.3.0` of the create with this change.

rfk commented:
Whadya know, we were actually jumping through a few unnecessary hoops ourselves in order to work around the API design here, and now that the top-level `legacy::*` helpers are available, we're able to make a nice cleanup in some of the consumer code that previously had to couple itself to a specific crypto backend. Ref https://github.com/mozilla/application-services/pull/3941/files#r598391958, and thanks again @tiesselune!
